cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/kafka/KafkaOffsetResolver.java

```diff
@@ -0,0 +1,316 @@
+   * same timestamp as in the given checkpoint. If they match, directly return {@code checkpoint.getNextOffset()}.
+   * If they don't, search for the smallest offset of the message with the same log event time
+   * as {@code checkpoint.getNextEventTime()}
+   *
+   * @param checkpoint A {@link Checkpoint} containing the next offset of a message and its log event timestamp.
+   * @param partition the partition in the topic for searching matching offset
+   * @return the next offset of the message with smallest offset and log event time equal to
+   *         {@code checkpoint.getNextEventTime()}.
+   *         {@code -1} if no such offset can be found or {@code checkpoint.getNextOffset()} is negative.
+   */
+   long getMatchingOffset(final Checkpoint checkpoint, final int partition) {
+    // Get BrokerInfo for constructing SimpleConsumer in OffsetFinderCallback
+    BrokerInfo brokerInfo = brokerService.getLeader(topic, partition);
+    if (brokerInfo == null) {
+      return kafka.api.OffsetRequest.EarliestTime(); // Return earliest offset if brokerInfo is null
+    }
+    SimpleConsumer simpleConsumer
+      = new SimpleConsumer(brokerInfo.getHost(), brokerInfo.getPort(),
+                           SO_TIMEOUT_MILLIS, FETCH_SIZE, "simple-kafka-client");
+
+    String clientId = topic + "_" + partition;
+    // Check whether the message fetched with the offset in the given checkpoint has the timestamp from
+    // checkpoint.getNextOffset() - 1 to get the offset corresponding to the timestamp in checkpoint
+    long timeStamp = getEventTimeByOffset(simpleConsumer, partition, clientId, checkpoint.getNextOffset() - 1);
+    if (timeStamp == checkpoint.getNextEventTime()) {
+      return checkpoint.getNextOffset();
+    }
+
+    long smallestMatchingOffset = findOffsetByTime(simpleConsumer, partition, clientId, checkpoint.getNextEventTime());
+    if (smallestMatchingOffset >= 0) {
+      // Increment smallestMatchingOffset to get the next offset if it's a valid non-negative offset
+      smallestMatchingOffset++;
+    }
+    return smallestMatchingOffset;
+  }
+
+  /**
+   * Determine the lower-bound and upper-bound of offsets to search for the smallest offset of the message
+   * with log event time equal to {@code targetTime}, and then perform the search.
+   *
+   * @return the matching offset or {@code -1} if not found
+   */
+  private long findOffsetByTime(SimpleConsumer consumer, int partition, String clientName, long targetTime) {
+    TopicAndPartition topicAndPartition = new TopicAndPartition(topic, partition);
+    Map<TopicAndPartition, PartitionOffsetRequestInfo> requestInfo = new HashMap<>();
+    // the offset of the earliest message
+    long minOffset = fetchOffsetByTime(consumer, partition, kafka.api.OffsetRequest.EarliestTime());
+    // the next offset after the latest message
+    long maxOffset = fetchOffsetByTime(consumer, partition, kafka.api.OffsetRequest.LatestTime());
```

Comment:
The bound is actually tighter. Let's say from container to Kafka, the publish delay is `d` and the replication delay is `r`, then for an event with timestamp `t`, the Kafka write time `kwt` should be `t <= kwt <= t + r + d`. An event can't be published before the event time and it won't be published after the assumed delay (or we don't care if that's the case).